### PR TITLE
(performance): speedup for `zarr`-based sparse indexing

### PR DESCRIPTION
**Pull request by ilan-gold**
```python
import scipy.sparse as sp
import numpy as np
import anndata as ad
import zarr
from pathlib import Path

data_path = Path('data/foo.zarr')
if not data_path.exists():
    z = zarr.open_group('data/foo.zarr', mode='w')
    arr = sp.random(1_000_000, 10_000, format="csr", random_state=np.random.default_rng())
    ad.io.write_elem(z, "X", arr)
z = zarr.open_group('data/foo.zarr', mode='r')

adata = ad.AnnData(X=ad.io.sparse_dataset(z["X"]))
idx = np.random.randint(0, adata.shape[0], 1024)
idx.sort()
%time adata[idx].X
```
On my computer, this is a 3-5X speedup.

- [ ] Closes #
- [ ] Tests added
- [x] Release note added (or unnecessary)

### CODE REVIEW

**Comment by ilan-gold** [comment]
Very good :)

File: .github/workflows/test-gpu.yml [hunk 1]
```diff
@@ -72,7 +72,9 @@ jobs:
           cache-dependency-path: pyproject.toml
 
       - name: Install AnnData
-        run: uv pip install --system -e ".[dev,test,cu12]" llvmlite>=0.43
+        run:
+          printf "llvmlite>=0.43\nscanpy>=1.10.0rc1" | tee /tmp/constraints.txt
+          uv pip install --system -e ".[dev,test,cu12]" -c /tmp/constraints.txt
```
Comment (ilan-gold) [inline .github/workflows/test-gpu.yml:77]: drive-by fix....

File: src/anndata/_core/sparse_dataset.py [hunk 2]
```diff
@@ -250,30 +250,44 @@ def slice_as_int(s: slice, l: int) -> int:
...
-    indices = np.concatenate([x.indices[s[0] : s[-1]] for s in indptr_sels])
+    indptr_list = [x.indptr[slice(s.start, s.stop + 1)] for s in slices]
+    # HDF5 cannot handle out-of-order integer indexing
+    if isinstance(x.data, ZarrArray):
+        indptr_int = np.concatenate([np.arange(s[0], s[-1]) for s in indptr_list])
+        data = x.data[indptr_int]
+        indices = x.indices[indptr_int]
+    else:
+        data = np.concatenate([x.data[s[0] : s[-1]] for s in indptr_list])
+        indices = np.concatenate([x.indices[s[0] : s[-1]] for s in indptr_list])
```
Comment (flying-sheep) [inline src/anndata/_core/sparse_dataset.py:279]:
```suggestion
    if isinstance(x.data, ZarrArray):
        indptr_int = np.concatenate([np.arange(start, end) for start, end in indptr_list])
        data = x.data[indptr_int]
        indices = x.indices[indptr_int]
    else:
        data = np.concatenate([x.data[start:end] for start, end in indptr_list])
        indices = np.concatenate([x.indices[start:end] for start, end in indptr_list])
```

File: src/anndata/_core/sparse_dataset.py [hunk 3]
```diff
@@ -250,30 +250,44 @@ def slice_as_int(s: slice, l: int) -> int:
...
     # Need to track the size of the gaps in the slices to each indptr subselection
-    total = indptr_sels[0][0]
+    total = indptr_list[0][0]
     offsets = [total]
-    for i, sel in enumerate(indptr_sels[1:]):
-        total = (sel[0] - indptr_sels[i][-1]) + total
+    for i, sel in enumerate(indptr_list[1:]):
+        total = (sel[0] - indptr_list[i][-1]) + total
```
Comment (flying-sheep) [inline src/anndata/_core/sparse_dataset.py:284]: Since you changed the name, `sel` is confusing here.

Also the whole loop is confusing. Please rewrite this to use `itertools.pairwise` instead of `enumerate`. (e.g. something like `for (s0, e0), (s1, e1) in pairwise(indptr_list): ...`)

And then use `sum(<generator comprehension>)` instead of a loop

**Review by flying-sheep — CHANGES_REQUESTED** [review]
great! I think for clarity, wherever possible you should unpack an entry of `indptr_list` into `start, end` instead of using `s[0]` and `s[1]`.

you could either do `indptr_limits = [(x.indptr[s.start], x.indptr[s.stop + 1]) for s in slices]` or so, or just define `indptr_slices` as above.

File: src/anndata/_core/sparse_dataset.py (hunk 2 above)
Comment (ilan-gold) [reply src/anndata/_core/sparse_dataset.py:279]: This doesn't work because each array in `indptr_limits` is not actually 2 elements, it could be more than that

File: src/anndata/_core/sparse_dataset.py (hunk 3 above)
Comment (ilan-gold) [reply src/anndata/_core/sparse_dataset.py:284]: Again, we run into the issue that `indptr_list` contains >2 element arrays.  I guess we just need two data structures here then.

File: src/anndata/_core/sparse_dataset.py (hunk 2 above)
Comment (flying-sheep) [reply src/anndata/_core/sparse_dataset.py]: please read what I wrote above, this refers to another change I asked for

Comment (ilan-gold) [reply src/anndata/_core/sparse_dataset.py]: Yup, I think I understood that you were suggesting two data structures.   I thought it was one in place of the other

File: src/anndata/_core/sparse_dataset.py [hunk 4]
```diff
@@ -250,30 +250,47 @@ def slice_as_int(s: slice, l: int) -> int:
...
     # Need to track the size of the gaps in the slices to each indptr subselection
-    total = indptr_sels[0][0]
-    offsets = [total]
-    for i, sel in enumerate(indptr_sels[1:]):
-        total = (sel[0] - indptr_sels[i][-1]) + total
-        offsets.append(total)
-    start_indptr = indptr_sels[0] - offsets[0]
+    offsets = [indptr_indices[0][0]]
+    for (_, stop0), (start1, _) in pairwise(indptr_limits):
+        offsets.append((start1 - stop0) + offsets[-1])
```
Comment (flying-sheep) [inline src/anndata/_core/sparse_dataset.py:288]: Should be something like this, right?

```suggestion
    gaps = (start1 - stop0 for (_, stop0), (start1, _) in pairwise(indptr_limits))
    offsets = accumulate(chain([indptr_limits[0][0]] gaps))
```

As said above, you could also do `indptr_slices` instead of `indptr_limits`, then we’d get

```diff
-indptr_int = np.concatenate([np.arange(start, stop) for start, stop in indptr_limits])
+indptr_int = np.concatenate([np.arange(s.start, s.stop) for s in indptr_slices])
```

and

```diff
-offsets = accumulate(chain([indptr_limits[0][0]] gaps))
+offsets = accumulate(chain([indptr_slices[0].start] gaps))
```

which I think is nicer

**Review by flying-sheep — APPROVED** [review]

File: src/anndata/_core/sparse_dataset.py (hunk 4 above)
Comment (ilan-gold) [reply src/anndata/_core/sparse_dataset.py:288]: The accumulation iterator is great.  It cleans up some code below as well.

**Review by flying-sheep — APPROVED** [review]
